enhancements/ingress/optional-ingress.md

```diff
@@ -0,0 +1,274 @@
+reviewers:
+  - "@Miciah"
+  - "@bparees"
+  - "@cewong"
+  - "@wking"
+approvers:
+  - "@Miciah"
+  - "@bparees"
+api-approvers:
+  - N/A
+creation-date: 2023-05-30
+last-updated: 2023-05-30
+tracking-link:
+  - https://issues.redhat.com/browse/NE-1129
+see-also:
+  - "/enhancements/installer/component-selection.md"
+  - "/enhancements/ingress/aws-load-balancer-operator.md"
+replaces:
+superseded-by:
+---
+
+# Make Ingress Operator optional
+
+## Summary
+
+This enhancement describes the implementation of a new capability to enable/disable the default ingress of the OpenShift cluster.
+The new capability (from here on called the ingress capability) includes [the cluster ingress operator](https://github.com/openshift/cluster-ingress-operator) as well as the custom resources managed by the operator.
+This enhancement focuses on the requirements for [RFE-3395](https://issues.redhat.com/browse/RFE-3395) which targets ROSA, a managed OpenShift service based on HyperShift. The implementation of the ingress capability on the standalone OpenShift is out of the scope of this enhancement.
+
+## Motivation
+
+Users of the ROSA managed OpenShift service would like to use [AWS load balancers](https://github.com/openshift/enhancements/blob/master/enhancements/ingress/aws-load-balancer-operator.md) as a way to accept the ingress traffic into the cluster.
+The cluster ingress operator becomes unnecessary and can be disabled to reduce the resource consumption.
+
+### Goals
+- Implement the new ingress capability.
+- Add the ingress capability to the ingress operator's payload.
+- Forbid the usage of the ingress capability for the standalone OpenShift.
+- Make the dependent cluster operators tolerate the absence of the ingress operator.
+
+### Non-Goals
+- Use a different way for enabling/disabling the ingress other than via [the cluster capability](https://github.com/openshift/enhancements/blob/master/enhancements/installer/component-selection.md).
+- Fully implement the ingress capability for the standalone OpenShift:
+  - Make the OpenShift installer tolerate the absence of the default ingress controller.
+  - Make the cluster authentication operator tolerate the absence of the default ingress controller.
+  - Make the cluster monitoring operator tolerate the absence of the default ingress controller.
+- Disable the Route API or try to tolerate the absence of the default ingress controller.
+- Implement the ingress operator disabling in the HyperShift's control plane operator.
+
+## Proposal
+
+### User Stories
+
+#### Disable the ingress on HyperShift
+
+The user of the HyperShift API wants to use an AWS ALB to route the traffic to its example application deployed on a hosted cluster.
+
+The user sets the capabilities in one of the HyperShift resources (**note**: the capability API is yet to be defined in HyperShift):
+```yaml
+apiVersion: hypershift.openshift.io/v1beta1
+kind: HostedCluster
+spec:
+  capabilities:
+    additionalDisabledCapabilities:
```

Comment:
From my summary of [the ClusterVersion `capabilities` discussion][1]:

> ### exclude
>
> The [enhancement] also provided a way to drop specific capabilities from the selected set (`inclusionDefault` or `baselineCapabilitySet`). Ben still feels like that will be popular, but David is skeptical, and because we can always add a property in this space later without breaking backwards compatibility, we're leaving it off for now.

Now that we have a few 4.y of ClusterVersion `capabilities` out in the wild, maybe there's some feedback on whether there is actually a demand for this approach (which would require working out what to do if a capability showed up in both the include and exclude lists)?  In the absence of that feedback, I'd expect we want to stick with the same `ClusterVersionCapabilitiesSpec` structure we use for ClusterVersion, just to avoid trying to map across a schema shift if/when we end up piping this down that far.

[1]: https://github.com/openshift/api/pull/1106/commits/5b82635ef101e7c10b5fcbbcfb81315bb7a0da20
[enhancement]: https://github.com/openshift/enhancements/blob/88cb7438f3ac0a8121e3cef87cb144097ece8cda/enhancements/installer/component-selection.md#implementation-detailsnotesconstraints-optional

Reply:
I agree that the HyperShift's API is better to be similar (or even the same) to the cluster version's, for the capability part. Just for the note, I didn't try to propose any update to the HyperShift's APIs. I just wanted to demonstrate the end to end workflow for disabling of the ingress. I put a note about the fact that this is a "not yet designed" API but may be it's not clear enough, let me rephrase it.

Reply:
I see you also added a "Design the capability API for HyperShift" non-goal.  It feels like there's a risk that the enhancement is accepted, but then this API design turns out to be harder than we'd thought.  But also, no worries from me if you want to defer the details of API discussion until after broader questions have been sorted out.  Feel free to mark this thread resolved.

Reply:
Yes I added this non goal as [a response to this thread](https://github.com/openshift/enhancements/pull/1415#discussion_r1223249242). I think it would be pragmatic to defer the discussions about the future HyperShift API. Also I don't feel like I'm the right person to propose the design for any HyperShift API.